File: app/system/actors/ExamAutoSaverActor.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package system.actors;

import controllers.SettingsController;
import impl.EmailComposer;
import io.ebean.DB;
import java.io.IOException;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import javax.inject.Inject;
import models.Exam;
import models.ExamEnrolment;
import models.ExamInspection;
import models.ExamParticipation;
import models.ExaminationEvent;
import models.GeneralSettings;
import models.Reservation;
import models.User;
import models.json.ExternalExam;
import org.apache.pekko.actor.AbstractActor;
import org.joda.time.DateTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import util.AppUtil;
import util.datetime.DateTimeHandler;

public class ExamAutoSaverActor extends AbstractActor {

    private final Logger logger = LoggerFactory.getLogger(ExamAutoSaverActor.class);

    private final EmailComposer composer;
    private final DateTimeHandler dateTimeHandler;

    @Inject
    public ExamAutoSaverActor(EmailComposer composer, DateTimeHandler dateTimeHandler) {
        this.composer = composer;
        this.dateTimeHandler = dateTimeHandler;
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
            .match(String.class, s -> {
                logger.debug("Starting check for ongoing exams ->");
                checkLocalExams();
                checkExternalExams();
                logger.debug("<- done");
            })
            .build();
    }

    private void checkLocalExams() {
        List<ExamParticipation> participations = DB.find(ExamParticipation.class)
            .fetch("exam")
            .fetch("reservation")
            .fetch("reservation.machine.room")
            .fetch("examinationEvent")
            .where()
            .isNull("ended")
            .or()
            .isNotNull("reservation")
            .isNotNull("examinationEvent")
            .endOr()
            .findList();

        if (participations.isEmpty()) {
            logger.debug("None found");
            return;
        }
        markEnded(participations);
    }

    private DateTime getNow(ExamParticipation participation) {
        if (participation.getExaminationEvent() != null) {
            return DateTime.now();
        }
        Reservation reservation = participation.getReservation();
        return dateTimeHandler.adjustDST(DateTime.now(), reservation.getMachine().getRoom());
    }

    private void markEnded(List<ExamParticipation> participations) {
        for (ExamParticipation participation : participations) {
            Exam exam = participation.getExam();
            Reservation reservation = participation.getReservation();
            ExaminationEvent event = participation.getExaminationEvent();
            DateTime reservationStart = new DateTime(reservation == null ? event.getStart() : reservation.getStartAt());
            DateTime participationTimeLimit = reservationStart.plusMinutes(exam.getDuration());
            DateTime now = getNow(participation);
            if (participationTimeLimit.isBefore(now)) {
                participation.setEnded(now);
                participation.setDuration(
                    new DateTime(participation.getEnded().getMillis() - participation.getStarted().getMillis())
                );

                GeneralSettings settings = SettingsController.getOrCreateSettings("review_deadline", null, "14");
                int deadlineDays = Integer.parseInt(settings.getValue());
                DateTime deadline = new DateTime(participation.getEnded()).plusDays(deadlineDays);
                participation.setDeadline(deadline);

                participation.save();
                logger.info("Setting exam {} state to REVIEW", exam.getId());
                exam.setState(Exam.State.REVIEW);
                exam.save();
                if (exam.isPrivate()) {
                    // Notify teachers
                    Set<User> recipients = new HashSet<>();
                    recipients.addAll(exam.getParent().getExamOwners());
                    recipients.addAll(
                        exam.getExamInspections().stream().map(ExamInspection::getUser).collect(Collectors.toSet())
                    );
                    AppUtil.notifyPrivateExamEnded(recipients, exam, composer);
                }
            } else {
                logger.info("Exam {} is ongoing until {}", exam.getId(), participationTimeLimit);
            }
        }
    }

    private void checkExternalExams() {
        List<ExamEnrolment> enrolments = DB.find(ExamEnrolment.class)
            .fetch("externalExam")
            .fetch("reservation")
            .fetch("reservation.machine.room")
            .where()
            .isNotNull("externalExam")
            .isNotNull("externalExam.started")
            .isNull("externalExam.finished")
            .isNotNull("reservation.externalRef")
            .findList();
        for (ExamEnrolment enrolment : enrolments) {
            ExternalExam exam = enrolment.getExternalExam();
            Exam content;
            try {
                content = exam.deserialize();
            } catch (IOException e) {
                logger.error("Failed to parse content out of an external exam", e);
                continue;
            }
            Reservation reservation = enrolment.getReservation();
            DateTime reservationStart = new DateTime(reservation.getStartAt());
            DateTime participationTimeLimit = reservationStart.plusMinutes(content.getDuration());
            DateTime now = dateTimeHandler.adjustDST(DateTime.now(), reservation.getMachine().getRoom());
            if (participationTimeLimit.isBefore(now)) {
                exam.setFinished(now);
                content.setState(Exam.State.REVIEW);
                try {
                    exam.serialize(content);
                    logger.info("Setting external exam {} state to REVIEW", exam.getHash());
                } catch (IOException e) {
                    logger.error("failed to parse content out of an external exam", e);
                }
            }
        }
    }
}

File: app/system/actors/ExamAutoSaverActor.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package system.actors

import controllers.admin.SettingsController
import impl.EmailComposer
import io.ebean.DB
import miscellaneous.datetime.DateTimeHandler
import miscellaneous.scala.DbApiHelper
import models.enrolment.{ExamEnrolment, ExamParticipation}
import models.exam.Exam
import org.apache.pekko.actor.AbstractActor
import org.joda.time.DateTime
import play.api.Logging

import java.io.IOException
import javax.inject.Inject
import scala.jdk.CollectionConverters.*
import scala.util.control.Exception.catching

class ExamAutoSaverActor @Inject (private val composer: EmailComposer, private val dateTimeHandler: DateTimeHandler)
    extends AbstractActor
    with Logging
    with DbApiHelper:

  override def createReceive(): AbstractActor.Receive = receiveBuilder()
    .`match`(
      classOf[String],
      (_: String) =>
        logger.debug("Starting check for ongoing exams ->")
        checkLocalExams()
        checkExternalExams()
        logger.debug("<- done")
    )
    .build

  private def checkLocalExams(): Unit =
    val participants = DB
      .find(classOf[ExamParticipation])
      .fetch("exam")
      .fetch("reservation")
      .fetch("reservation.machine.room")
      .fetch("examinationEvent")
      .where
      .isNull("ended")
      .or
      .isNotNull("reservation")
      .isNotNull("examinationEvent")
      .endOr
      .list
    if participants.isEmpty then logger.debug("None found")
    else markEnded(participants)

  private def getNow(participation: ExamParticipation) =
    if Option(participation.getExaminationEvent).nonEmpty then DateTime.now
    else
      val reservation = participation.getReservation
      dateTimeHandler.adjustDST(DateTime.now, reservation.getMachine.getRoom)

  private def markEnded(participants: List[ExamParticipation]): Unit =
    participants.foreach(participation =>
      val exam        = participation.getExam
      val reservation = participation.getReservation
      val event       = participation.getExaminationEvent
      val reservationStart = new DateTime(
        if Option(reservation).isEmpty then event.getStart
        else reservation.getStartAt
      )
      val participationTimeLimit = reservationStart.plusMinutes(exam.getDuration)
      val now                    = getNow(participation)
      if participationTimeLimit.isBefore(now) then
        participation.setEnded(now)
        participation.setDuration(new DateTime(participation.getEnded.getMillis - participation.getStarted.getMillis))
        val settings     = SettingsController.getOrCreateSettings("review_deadline", null, "14")
        val deadlineDays = settings.getValue.toInt
        val deadline     = new DateTime(participation.getEnded).plusDays(deadlineDays)
        participation.setDeadline(deadline)
        participation.save()
        logger.info(s"Setting exam ${exam.getId} state to REVIEW")
        exam.save()
        if exam.isPrivate then
          // Notify teachers
          val recipients = exam.getParent.getExamOwners.asScala ++ exam.getExamInspections.asScala.map(_.getUser)
          recipients.foreach(r =>
            composer.composePrivateExamEnded(r, exam)
            logger.info(s"Email sent to ${r.getEmail}")
          )
      else logger.info(s"Exam ${exam.getId} is ongoing until $participationTimeLimit")
    )

  private def checkExternalExams(): Unit =
    DB
      .find(classOf[ExamEnrolment])
      .fetch("externalExam")
      .fetch("reservation")
      .fetch("reservation.machine.room")
      .where
      .isNotNull("externalExam")
      .isNotNull("externalExam.started")
      .isNull("externalExam.finished")
      .isNotNull("reservation.externalRef")
      .list
      .flatMap(enrolment =>
        catching(classOf[IOException]).either(enrolment.getExternalExam.deserialize()) match
          case Left(e) =>
            logger.error("Failed to parse content out of an external exam", e)
            None
          case Right(content) => Some((enrolment, content))
      )
      .foreach((enrolment, content) =>
        val (exam, reservation)    = (enrolment.getExternalExam, enrolment.getReservation)
        val reservationStart       = new DateTime(reservation.getStartAt)
        val participationTimeLimit = reservationStart.plusMinutes(content.getDuration)
        val now                    = dateTimeHandler.adjustDST(DateTime.now, reservation.getMachine.getRoom)
        if participationTimeLimit.isBefore(now) then
          exam.setFinished(now)
          content.setState(Exam.State.REVIEW)
          catching(classOf[IOException]).either(exam.serialize(content)) match
            case Left(e)  => logger.error("failed to parse content out of an external exam", e)
            case Right(_) => logger.info(s"Setting external exam ${exam.getHash} state to REVIEW")
      )
